server/src/ankaios_server/cyclic_check.rs

```diff
@@ -0,0 +1,634 @@
+// Copyright (c) 2024 Elektrobit Automotive GmbH
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache License, Version 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0.
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+use common::objects::State;
+use std::collections::{HashSet, VecDeque};
+
+/// Returns an Option containing the workload dependency that is part of a cycle
+/// or [`None`] if no cycles are detected
+///
+/// The method implements the iterative depth search first (dfs) algorithm to
+/// detect a cycle in the directed graph for inter workload dependencies within a state.
+///
+/// # Arguments
+///
+/// * `state` - The State with workloads representing the directed graph to check for a cycle
+/// * `start_nodes` - Start visiting the graph only for the passed workloads
+///                   if [`None`] the search is started from all workloads of the state
+///
+pub fn dfs(state: &State, start_nodes: Option<Vec<&String>>) -> Option<String> {
+    // stack is used to terminate the search properly
+    let mut stack: VecDeque<&String> = VecDeque::new();
+
+    // used to prevent visiting nodes repeatedly
+    let mut visited: HashSet<&String> = HashSet::with_capacity(state.workloads.len());
+
+    /* although the path container is used for lookups,
+    measurements have shown that it is faster than associative data structure within this code path */
+    let mut path: VecDeque<&String> = VecDeque::with_capacity(state.workloads.len());
+
+    // start visiting workloads in the graph only for a subset of workloads (e.g. in case of a an update) or for all
+    let mut data: Vec<&String> = if let Some(workloads_to_visit) = start_nodes {
```

Comment:
Well, I do not like name of the variable `data`, but I do not have better name for it.